common/predictive-text/README.md

```diff
@@ -0,0 +1,404 @@
+ 3. Tokens **MUST** be unique across messages. That is, tokens **MUST
+    NOT** be duplicated between different messages.
+
+It is up to the keyboard to create unambiguous tokens that can be uniquely
+identified through the round-trip process.
+
+In the following examples, I'll use the subset of `number` values that
+are interpretable as [31-bit signed integers][Smi].
+
+[Map object]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#Key_equality
+[Smi]: https://github.com/thlorenz/v8-perf/blob/master/data-types.md#efficiently-representing-values-and-tagging
+[Tokens]: #tokens
+
+### Example
+
+An asynchronous message to predict after typing 'D':
+
+```javascript
+{
+    message: 'predict',
+
+    token: 1,
+    transform: {
+        insert: 'D',
+        deleteLeft: 0,
+        deleteRight: 0
+    },
+    context: {
+      left: '',
+      right: '',
+      startOfBuffer: true,
+      endOfBuffer: true
+    },
+}
+```
+
+Message types
+-------------
+
+Currently there are four message types:
+
+Message       | Direction          | Parameters          | Expected reply      | Uses token
+--------------|--------------------|---------------------|---------------------|---------------
+`initialize`  | keyboard → LMLayer | initialization      | Yes — `ready`       | No
+`ready`       | LMLayer → keyboard | configuration       | No                  | No
+`predict`     | keyboard → LMLayer | transform, context  | Yes — `suggestions` | Yes
+`suggestions` | LMLayer → keyboard | suggestions         | No                  | Yes
+
+
+### Message: `initialize`
+
+Must be sent from the keyboard to the LMLayer so that the LMLayer
+initializes a model. It will send `initialization` which is a plain
+JavaScript object specify the path to the model, as well configurations
+and platform restrictions.
+
+The keyboard **MUST NOT** send any messages to the LMLayer prior to
+sending `initialize`. The keyboard **SHOULD NOT** send another message
+to the keyboard until it receives `ready` message from the LMLayer
+before sending another message.
+
+These are the configurations, and platform restrictions sent to
+initialize the LMLayer and its model.
+
+```typescript
+interface InitializeMessage {
+  message: 'initialize';
+
+  /**
+   * Path to the model. There are no concrete restrictions on the path
+   * to the model, so long as the LMLayer can successfully use it to
+   * initialize the model.
+   */
+  model: string;
+
+  configuration: {
+    /**
+     * Whether the platform supports right contexts.
+     * The absence of this rule implies false.
+     */
+    supportsRightContexts?: false,
+
+    /**
+     * Whether the platform supports deleting to the right.
+     * The absence of this rule implies false.
+     */
+    supportsDeleteRight?: false,
+
+    /**
+     * The maximum amount of UTF-16 code units that the keyboard will
+     * provide to the left of the cursor.
+     */
+    maxLeftContextCodeUnits: 32,
```

Comment:
I assume this is an example implementation of the object, rather than its spec?  I noticed that the `ReadyMessage` object following this one uses types instead of literals for its field declarations.

Reply:
Correct. I think I just forgot to convert this one from a concrete implementation